src/pal/src/misc/miscpalapi.cpp

```diff
@@ -226,86 +234,159 @@ PAL_GetPALDirectoryA(
     return bRet;
 }
 
-VOID
-PALAPI
-PAL_Random(
+static
+BOOL
+PAL_RandomFromUrandom(
         IN OUT LPVOID lpBuffer,
         IN DWORD dwLength)
 {
-    int rand_des = -1;
-    DWORD i;
-    long num = 0;
     static BOOL sMissingDevURandom;
-    static BOOL sInitializedMRand;
+    int rand_des;
 
-    PERF_ENTRY(PAL_Random);
-    ENTRY("PAL_Random(lpBuffer=%p, dwLength=%d)\n", lpBuffer, dwLength);
+    if (sMissingDevURandom)
+    {
+        return FALSE;
+    }
 
-    if (!sMissingDevURandom)
+    do
     {
-        do
+        rand_des = open(URANDOM_DEVICE_NAME, O_RDONLY | O_CLOEXEC);
+    } while ((rand_des == -1) && (errno == EINTR));
+
+    if (rand_des == -1)
+    {
+        if (errno == ENOENT)
         {
-            rand_des = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
+            sMissingDevURandom = TRUE;
         }
-        while ((rand_des == -1) && (errno == EINTR));
+        else
+        {
+            ASSERT("PAL__open() failed, errno:%d (%s)\n", errno, strerror(errno));
+        }
+
+        return FALSE;
+    }
 
-        if (rand_des == -1)
+    DWORD offset = 0;
+    do
+    {
+        ssize_t n = read(rand_des, (BYTE*)lpBuffer + offset , dwLength - offset);
+        if (n == -1)
         {
-            if (errno == ENOENT)
+            if (errno == EINTR)
             {
-                sMissingDevURandom = TRUE;
-            }
-            else
-            {
-                ASSERT("PAL__open() failed, errno:%d (%s)\n", errno, strerror(errno));
+                continue;
             }
+            ASSERT("read() failed, errno:%d (%s)\n", errno, strerror(errno));
 
-            // Back off and try mrand48.
+            break;
         }
-        else
+
+        offset += n;
+    }
+    while (offset != dwLength);
+
+    _ASSERTE(offset == dwLength);
+
+    close(rand_des);
+
+    return TRUE;
+}
+
+#if defined(HAVE_SYS_GETRANDOM)
```

Comment:
This needs to be `#if HAVE_SYS_GETRANDOM`, cmake always define the symbol, it is defined as 0 when the test failed and 1 when it succeeded.

Reply:
Done